### PR TITLE
Document try!'s error conversion behaviour

### PR DESCRIPTION
**Pull request by skade**
try!'s documentation currently doesn't document the error conversion behaviour of the macro. This patch extends the documentation.

Open questions:
- is it worthwhile to have seperate examples with and without wrapping behaviour? It's not immediately obvious that From<T> for T is always defined. Though this is necessary for the macro to work in any case, is this the place to expect that knowledge.

### CODE REVIEW

File: src/libcore/macros.rs [hunk 1]
```diff
@@ -190,9 +190,18 @@ macro_rules! debug_assert_eq {
 }
 
 /// Helper macro for unwrapping `Result` values while returning early with an
-/// error if the value of the expression is `Err`. Can only be used in
-/// functions that return `Result` because of the early return of `Err` that
-/// it provides.
+/// error if the value of the expression is `Err`. It additionally provides
+/// automatic conversions of errors into other errors using `From`.
+/// Can only be used in functions that return `Result` because of the
```
Comment (apasel422) [inline src/libcore/macros.rs]: s/Can/It can/

File: src/libcore/macros.rs [hunk 2]
```diff
@@ -190,9 +190,18 @@ macro_rules! debug_assert_eq {
...
+/// early return of `Err` that it provides.
+///
+/// `try!` unwraps the given `Result` and, in case of an `Ok` variant, the
+/// expression has the value of the wrapped value.
+///
+/// In case of an `Err` variant, it unwraps the inner error. It then performs
+/// conversion using `From`. This provides automatic conversion between two
+/// specialized errors and more general ones. It then returns the resulting
+/// error is then immediately returned.
```
Comment (apasel422) [inline src/libcore/macros.rs]: Look like there's an extra "is then immediately returned" here.

File: src/libcore/macros.rs [hunk 3]
```diff
@@ -190,9 +190,18 @@ macro_rules! debug_assert_eq {
...
 /// Helper macro for unwrapping `Result` values while returning early with an
-/// error if the value of the expression is `Err`. Can only be used in
-/// functions that return `Result` because of the early return of `Err` that
-/// it provides.
+/// error if the value of the expression is `Err`. It additionally provides
+/// automatic conversions of errors into other errors using `From`.
+/// Can only be used in functions that return `Result` because of the
+/// early return of `Err` that it provides.
+///
+/// `try!` unwraps the given `Result` and, in case of an `Ok` variant, the
```
Comment (steveklabnik) [inline src/libcore/macros.rs]: "unwrap" is an unfortunate choice of words here, since `unwrap()` is a method, and it doesn't use it

**Comment by skade** [comment]
Fixed all comments.

**Comment by skade** [comment]
@steveklabnik updated. Made it more concise (hopefully) and direct.

**Comment by steveklabnik** [comment]
I'm into it.  Build failure looks spurious. Thanks!

@bors: r+ rollup

**Pull request merged**